File: src/main/java/org/springframework/samples/petclinic/rest/dto/PagedResultDto.java
package org.springframework.samples.petclinic.rest.dto;

import java.util.List;
import java.util.function.Function;

import org.springframework.data.domain.Page;

public record PagedResultDto<T>(List<T> content, int pageNumber, int pageSize, long totalElements, int totalPages) {

	public static <E, D> PagedResultDto<D> from(Page<E> page, Function<E, D> mapper) {
		return new PagedResultDto<>(page.getContent().stream().map(mapper).toList(), page.getNumber() + 1,
				page.getSize(), page.getTotalElements(), page.getTotalPages());
	}

}

File: src/main/kotlin/org/springframework/samples/petclinic/rest/dto/PagedResultDto.kt
package org.springframework.samples.petclinic.rest.dto

import org.springframework.data.domain.Page

data class PagedResultDto<T : Any>(
    val content: List<T>,
    val pageNumber: Int,
    val pageSize: Int,
    val totalElements: Long,
    val totalPages: Int
) {
    companion object {
        fun <E : Any, D : Any> from(page: Page<E>, mapper: (E) -> D): PagedResultDto<D> {
            return PagedResultDto(
                content = page.content.map(mapper),
                pageNumber = page.number + 1,
                pageSize = page.size,
                totalElements = page.totalElements,
                totalPages = page.totalPages
            )
        }
    }
}